src/DebugEngineHost/HostNatvisProject.cs

```diff
@@ -73,6 +81,74 @@ internal enum VSENUMPROJFLAGS
                 //  file as well as all virtual projects including nested (a.k.a. sub) projects
             };
 
+            /// <summary>
+            /// Gets the WorkspaceService from Microsoft.VisualStudio.Workspace
+            /// </summary>
+            /// <remarks>This package won't be automatically loaded by MEF, so we need to manually acquire exported MEF Parts.</remarks>
+            private static IVsFolderWorkspaceService GetWorkspaceService()
+            {
+                IComponentModel componentModel = ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel).GUID) as IComponentModel;
+                var workspaceServices = componentModel.DefaultExportProvider.GetExports<IVsFolderWorkspaceService>();
+
+                if (workspaceServices != null && workspaceServices.Count() > 0)
+                {
+                    return workspaceServices.First().Value;
+                }
+                return null;
+            }
+
+            private static int GetOpenFolderSourceLocations(string bstrFileName, out bool isIndexComplete, out IEnumerable<string> pSourcesArray)
+            {
+                isIndexComplete = false;
+                var workspaceService = GetWorkspaceService();
+
+                if (workspaceService != null)
+                {
+                    IWorkspace currentWorkspace = workspaceService.CurrentWorkspace;
+                    IIndexWorkspaceService indexWorkspaceService = currentWorkspace?.GetService<IIndexWorkspaceService>(throwIfNotFound: false);
+                    if (indexWorkspaceService != null)
+                    {
+                        isIndexComplete = indexWorkspaceService.State == IndexWorkspaceState.Completed;
+
+                        var findFilesService = indexWorkspaceService as IFindFilesService;
+                        if (findFilesService != null)
+                        {
+                            FindFileServiceProgress progress = new FindFileServiceProgress();
+                            string toSearch = Path.GetFileName(bstrFileName);
+                            using (var cancellationTokenSource = new CancellationTokenSource())
+                            {
+                                ThreadHelper.JoinableTaskFactory.Run(async () =>
```

Comment:
It feels like we should be making this function async instead so that this code can run off the UI thread.